### PR TITLE
Add InitialData base class for advection system

### PR DESCRIPTION
**Pull request by yoonso0-0**
## Proposed changes

 - Add Initial Data group to Groupdefs.
 - Define an `InitialData` abstract base class in `PointwiseFunctions/InitialData.hpp` which has analytic solutions and analytic data as its derived classes. Related files are in `PointwiseFunctions/Tags/` subdirectory.  A derived class named `NumericInitialDataFromFile`, for importing numerc initial data, will be added in the upcoming PRs.
 - For the ScalarAdvection system, add the `InitialData` base class to factory creation list and its tag to global cache tags. The goal is to make initial data specifiable via input file like:
  ```
  InitialData:
    Sinusoid:  # or Krivodonova, Kuzmin, .. (other types of solutions or data)
      # Options for each analytic solution / analytic data
  
  InitialData:
    NumericInitialDataFromFile:
  ```
- Derive analytic solutions and analytic data of advection system from the new base class. Enable the code to specify the kind of initial data via input file, without compiling separate executables for different types of initial data. Remove `InitialData` template parameter from evolution metavariables and cleanup.

### Upgrade instructions
- All available analytic data (solutions) are listed in `PointwiseFunctions/AnalyticData/ScalarAdvection/Factory.hpp` (`PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Factory.hpp`) with a templated type alias named `all_analytic_data` (`all_analytic_solutions`).  Rather than listing all the types of analytic data or solution and their header files, it is sufficient to just include two Factory.hpp and add the type aliases to `factory classes` in EvolutionMetavars.
- We have `EvolveScalarAdvection<DIM>` (where DIM = 1D or 2D) target executables and can specify initial data with the option `InitialData:` in the input files.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp
Comment (nilsvu) [inline src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp]: Name this just `InitialData` please. You can make the default-constructor `protected` if you want to make sure the class is abstract (can't be constructed).

File: src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp
Comment (nilsvu) [inline src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp]: Delete this class please, since it's just a placeholder. I think you can derive the `Sinusoid` solution from `InitialData` already, without breaking any existing code.

File: src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp
Comment (nilsvu) [inline src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp]: You can just name this `InitialData`, and then you can delete the `name()` function.

File: src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp
Comment (nilsvu) [inline src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp]: My suggestion is to move this tag and the option tag to a new file `Evolution/Tags/InitialData.hpp`. Then template both on `typename InitialDataType` and use it in the `std::unique_ptr`. Finally, add the following:

```cpp
namespace evolution::Tags {
struct InitialDataBase : db::BaseTag {};

template <typename InitialDataType>
struct InitialData : InitialDataBase, db::SimpleTag {
  using type = std::unique_ptr<InitialDataType>;
  /// ...
```

With this setup you can use `db::get<evolution::Tags::InitialDataBase>(box)` to retrieve the initial-data object without knowing its type. This will be useful to write code that's not specific to a particular system.

**Comment by wthrowe** [comment]
I think we should wait to add the class until it has some sort of functionality.

**Comment by yoonso0-0** [comment]
@wthrowe I'll mark this PR as 'in progress' until there comes a practical change in functionality. How about that?

**Review by nilsvu — CHANGES_REQUESTED** [review]
Looks good! You can squash these changes directly, and feel free to push more commits to this branch.

**Comment by yoonso0-0** [comment]
@wthrowe Yes. I marked this PR as 'in progress' -- just wanted to get feedbacks and make sure there's nothing _really_ bad about this base class, before piling up other commits.

File: src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp
Comment (yoonso0-0) [reply src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp]: @nilsleiffischer How the global cache tag (which was `Tags::InitialData` beforehand) needs to be modified after this change? We need a template parameter for the `evolution::Tags::InitialData<#>` (the new one), but we are trying to avoid using the ```using initial_data = InitialData``` provided as a template paramter of current version EvolutionMetavars.

File: src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp
Comment (yoonso0-0) [reply src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp]: I deleted this class in the current PR.

File: src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp
Comment (nilsvu) [reply src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp]: In the metavariables you add the tag to the `const_global_cache_tags` like this:

```
evolution::Tags::InitialData<ScalarAdvection::InitialData::InitialData>
```

Note you pass the base class here, not a particular derived class. If at some point we find that we can share the `InitialData` base class between systems then the template parameter can be removed, but that's an easy change.

**Comment by nilsvu** [comment]
@yoonso0-0 My general advice for how to proceed is: Don't worry about numeric initial data for now. First derive the `Sinusoid` class from `InitialData` so it can be factory-created. Then replace the `SetVariables` action with a new one that traverses the `factory_classes` and calls their `variables` function to get the initial data. That needs to use our `call_with_dynamic_type` mechanism which is probably hard to get right on the first try, so feel free to message me so we can go over it together once you're at that point.

**Comment by yoonso0-0** [comment]
@nilsleiffischer Pushed up to today's call. I think you can skip the cosine data commit (it's mostly a copy&paste from sinusoid solution, and also works fine). There are a number of lines marked temporary or 'this needs to be done' -- feel free to suggest any changes to make these look less cluttered.

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp
Comment (nilsvu) [inline src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp]: I think it's better to delete this alias and put the tag in the `const_global_cache_tags` list directly, so just drop this change from the commit.

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp
Comment (nilsvu) [inline src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp]: thanks for the comment, but from my side it can be deleted :)

File: src/Evolution/Initialization/InitialData.hpp
Comment (nilsvu) [inline src/Evolution/Initialization/InitialData.hpp]: delete, and the `<iostream>` include

File: src/Evolution/Initialization/InitialData.hpp
Comment (nilsvu) [inline src/Evolution/Initialization/InitialData.hpp]: You can make the return type explicit: `tuples::TaggedTuple<Tags...>`

File: src/Evolution/Initialization/InitialData.hpp
Comment (nilsvu) [inline src/Evolution/Initialization/InitialData.hpp]: remove the namespace. I think the new function shouldn't collide with the existing ones, but not entirely sure.

File: src/Evolution/Initialization/SetVariables.hpp
Comment (nilsvu) [inline src/Evolution/Initialization/SetVariables.hpp]: delete

File: src/Evolution/Initialization/SetVariables.hpp
Comment (nilsvu) [inline src/Evolution/Initialization/SetVariables.hpp]: How about `SetInitialData`

**Review by nilsvu — CHANGES_REQUESTED** [review]
Looks good, mostly just the obvious "remove the temporary comments"

**Comment by yoonso0-0** [comment]
@nilsleiffischer All requested changes updated so far :) The new function did not make conflicts.

**Comment by nilsvu** [comment]
Nice! I think the only thing that's left is to handle observations of the error to the analytic solution. Now we only know at runtime if we have an analytic solution or not, so we have to store it in a `std::optional`. You can use the existing tag `AnalyticSolutionsOptional` in `PointwiseFunctions/AnalyticSolutions/Tags.hpp` for this. I suggest you create a new compute item for this tag and replace `evolution::Tags::AnalyticCompute` with it. Let me know in case you get stuck.

File: src/Evolution/Initialization/InitialData.hpp
Comment (yoonso0-0) [inline src/Evolution/Initialization/InitialData.hpp:52]: @nilsleiffischer If I drop this `constexpr` from if branch, I get the following message:
```
error: no matching member function for call to 'variables'
          return initial_data_derived->variables(
```
what's happening here? Is it related with that we are using `auto*` at the line 43?

**Review by kidder — CHANGES_REQUESTED** [review]
I am requesting a change as I would like to review this PR.

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp
Comment (yoonso0-0) [inline src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp]: This line needs to be reviewed -- should we place `::Tags::AnalyticSolutionsOptional<observe_fields>` instead..?

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp
Comment (nilsvu) [inline src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp]: I think you can just delete the `Cosine` class now since you added it for testing, right?

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp
Comment (nilsvu) [inline src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp]: My suggestion is you split this list and move it to `PointwiseFunctions/Analytic{Solutions,Data}/ScalarAdvection/Factory.hpp`. You'll have to template it on `Dim`. Then do something like this in the `factory_classes` below:

```cpp
tmpl::pair<InitialData, tmpl::append<
  ScalarAdvection::AnalyticSolutions::all_analytic_solutions<Dim>,
  ScalarAdvection::AnalyticData::all_analytic_data<Dim>>>
```

You won't have to `include` all the analytic classes here in the executable anymore, just the two `Factory.hpp` files.

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp
Comment (nilsvu) [inline src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp:126]: You can just keep this alias and set it equal to `observe_fields`. In other executables we may want to me more selective about which fields to observe and which we compute errors for, so the alias is a handy way to specify this.

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp
Comment (nilsvu) [reply src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp]: Drop this particular change (see comment above)

File: src/Evolution/Initialization/InitialData.hpp
Comment (nilsvu) [reply src/Evolution/Initialization/InitialData.hpp:52]: Yes it is. The lambda gets instantiated once for every derived class in the list, and the `auto` gets inferred to be that type. Then the `constexpr` means the condition is evaluated at compile-time, so in every instantiation of the lambda the compiler only sees one of the branches of the if-clause. If you remove the `constexpr` then the condition only gets evaluated at runtime, so the compiler tries to compile both branches for every derived class. That fails because analytic data classes have no `variables` function with a `time` argument. Let me know if anything is still unclear.

File: src/Evolution/Initialization/SetVariables.hpp
Comment (nilsvu) [inline src/Evolution/Initialization/SetVariables.hpp]: Could you move this action into a separate file please?

File: src/Evolution/Initialization/SetVariables.hpp
Comment (nilsvu) [inline src/Evolution/Initialization/SetVariables.hpp]: Just `initialization_tags`, so this gets picked up by the code that assembles options. You can fix the name in the other action as well, or leave it, up to you.

File: src/Evolution/Initialization/SetVariables.hpp
Comment (nilsvu) [inline src/Evolution/Initialization/SetVariables.hpp]: I think `std::move(inertial_coords)` because it's computed above (not only referenced) and can be forwarded here.

File: src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp
Comment (nilsvu) [inline src/Evolution/Systems/ScalarAdvection/InitialData/InitialData.hpp]: I think you can just move this to `Evolution/InitialData.hpp` and into the `evolution` namespace, and we'll use it for all systems. We can still add subclasses for specific systems if we need them.

File: src/Evolution/Tags/ComputeTags.hpp
Comment (nilsvu) [inline src/Evolution/Tags/ComputeTags.hpp]: Name this `AnalyticSolutionOptionalCompute`, in line with the naming of other compute tags.

File: src/Evolution/Tags/ComputeTags.hpp
Comment (nilsvu) [inline src/Evolution/Tags/ComputeTags.hpp]: Please name this file `AnalyticSolutionCompute.hpp` (we have started to put tags into individual files if they are likely to be included separately)

File: src/PointwiseFunctions/AnalyticData/ScalarAdvection/Cosine.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/ScalarAdvection/Cosine.hpp]: Can you perhaps think of an analytic data class that would be more useful than the Cosine?

**Review by nilsvu — CHANGES_REQUESTED** [review]
Please feel free to squash things directly

**Comment by nilsvu** [comment]
I think the things you need to add tests for are: the `SetInitialData` action and the compute tag. But it might be more efficient to wait for a coarse second review before writing those tests.

@kidder I think it would be reasonable to take a look at this point, or perhaps once @yoonso0-0 has squashed in the changes above.

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp
Comment (yoonso0-0) [reply src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp:126]: Got it.

File: src/Evolution/Initialization/SetVariables.hpp
Comment (yoonso0-0) [reply src/Evolution/Initialization/SetVariables.hpp]: Exactly what kind of change would this make on executables?

File: src/PointwiseFunctions/AnalyticData/ScalarAdvection/Cosine.hpp
Comment (yoonso0-0) [reply src/PointwiseFunctions/AnalyticData/ScalarAdvection/Cosine.hpp]: Nothing really 'more useful' -- this is very simple system and analytic solution is given quite trivially for every analytic data, so there is no analytic data which does not have solution. Maybe we can try with a profile a little bit more complicated than cosine function? i.e. half-circle or step.

File: src/Evolution/Initialization/SetVariables.hpp
Comment (nilsvu) [reply src/Evolution/Initialization/SetVariables.hpp]: Doesn't change anything because other actions list the InitiaTime Tag already. Currently the misspelled alias is just ignored.

File: src/PointwiseFunctions/AnalyticData/ScalarAdvection/Cosine.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticData/ScalarAdvection/Cosine.hpp]: Ok, no need to have analytic data classes for the advection system then. The action and compute tag are tested independently of the system anyway.

**Comment by yoonso0-0** [comment]
@nilsleiffischer @kidder I squashed the proposed changes and pushed them after doing some rearrangements.

First two commits are minor fixes. The third one (adding cosine initial data) is not something really necessary, but for providing at least one example of how to handle analytic data classes. The commits with `[wip]` are without corresponding tests yet.

**Comment by nilsvu** [comment]
Looks good! I think you can just remove all things related to analytic data from the scalar advection system, and squash the "Use factory header" commit into the one before. Then I think this is ready to go once the action and compute tag have tests 👍

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp
Comment (nilsvu) [inline src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp]: This `tmpl::list` isn't needed

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Factory.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Factory.hpp]: Wrap this in `tmpl::flatten`

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Comment by nilsvu** [comment]
@kidder this is ready for your review

**Comment by yoonso0-0** [comment]
I'm closing this PR since my fork branch is at working state and there is no code here to be merged.

**Pull request closed without merging**
